Source/WebCore/Modules/webcodecs/WebCodecsAudioData.cpp

```diff
@@ -0,0 +1,169 @@
+        });
+    }
+}
+
+std::optional<AudioSampleFormat> WebCodecsAudioData::format() const
+{
+    return m_data.audioData ? std::make_optional(m_data.audioData->format()) : std::nullopt;
+}
+
+float WebCodecsAudioData::sampleRate() const
+{
+    return m_data.audioData ? m_data.audioData->sampleRate() : 0;
+}
+
+size_t WebCodecsAudioData::numberOfFrames() const
+{
+    return m_data.audioData ? m_data.audioData->numberOfFrames() : 0;
+}
+
+size_t WebCodecsAudioData::numberOfChannels() const
+{
+    return m_data.audioData ? m_data.audioData->numberOfChannels() : 0;
+}
+
+std::optional<uint64_t> WebCodecsAudioData::duration()
+{
+    return m_data.audioData ? m_data.audioData->duration() : std::nullopt;
+}
+
+int64_t WebCodecsAudioData::timestamp() const
+{
+    return m_data.audioData ? m_data.audioData->timestamp() : 0;
+}
+
+// https://www.w3.org/TR/webcodecs/#dom-audiodata-allocationsize
+ExceptionOr<size_t> WebCodecsAudioData::allocationSize(const CopyToOptions& options)
+{
+    if (isDetached())
+        return Exception { InvalidStateError, "AudioData is detached"_s };
+
+    auto copyElementCount = computeCopyElementCount(*this, options);
+    if (copyElementCount.hasException())
+        return copyElementCount.releaseException();
+
+    auto destFormat = options.format.value_or(*format());
+    auto bytesPerSample = computeBytesPerSample(destFormat);
+    return copyElementCount.releaseReturnValue() * bytesPerSample;
+}
+
+// https://www.w3.org/TR/webcodecs/#dom-audiodata-copyto
+ExceptionOr<void> WebCodecsAudioData::copyTo(BufferSource&& source, CopyToOptions&& options)
+{
+    if (isDetached())
+        return Exception { InvalidStateError, "AudioData is detached"_s };
+
+    auto copyElementCount = computeCopyElementCount(*this, options);
+    if (copyElementCount.hasException())
+        return copyElementCount.releaseException();
+
+    auto destFormat = options.format.value_or(*format());
+    auto bytesPerSample = computeBytesPerSample(destFormat);
+    auto maxCopyElementCount = copyElementCount.releaseReturnValue();
+    auto allocationSize = maxCopyElementCount * bytesPerSample;
+
+    if (allocationSize > source.length())
+        return Exception { RangeError, "Buffer is too small"_s };
+
+    std::span<uint8_t> buffer { static_cast<uint8_t*>(source.mutableData()), source.length() };
+    m_data.audioData->copyTo(buffer, destFormat, options.planeIndex, options.frameOffset, options.frameCount, maxCopyElementCount);
+    return { };
+}
+
+// https://www.w3.org/TR/webcodecs/#dom-audiodata-clone
+ExceptionOr<Ref<WebCodecsAudioData>> WebCodecsAudioData::clone(ScriptExecutionContext& context)
+{
+    if (isDetached())
+        return Exception { InvalidStateError,  "AudioData is detached"_s };
+
+    auto clone = adoptRef(*new WebCodecsAudioData(context, WebCodecsAudioInternalData { m_data }));
+
+    clone->m_isDetached = m_isDetached;
```

Comment:
Do we need this given we reject when AudioData is detached?

Reply:
No need indeed.